[leehuun][react-week5-assignment-1]

huun0424 opened the pull request:
안녕하세요 트레이너님.
금일 작업 내용 PR 올립니다.
전주, 금주 강의 내용 복기하면서 구조 잡느라 많은 내용을 하진 못했습니다.
reducer, action 구조 잡는 정도로 진행했고,
내일부터 test 더 보완하면서 comp 추가 작업 진행하겠습니다. :)

작업내용.
 - Regions comp, test file 추가.
 - RegionsContainer comp 추가.
 - reducer 설계.
 - store 추가.
 - action 추가.

Commit:
[leehuun][react-week5-assignment-1]

 - Regions comp, test file 추가.
 - RegionsContainer comp 추가.
 - reducer 설계.
 - store 추가.
 - action 추가.

passwd10 commented on src/App.test.jsx:
```diff
@@ -0,0 +1,26 @@
+
+jest.mock('react-redux');
+
+it('App', () => {
```
`describe` 로 App 테스트를 표현해주시고, App 에 대한 각각의 테스트는 `it` 으로 작성해주세요!

```javascript
describe('App', () => {
  it('~~를 렌더링한다') => {
    ...
  }
})

```

passwd10 commented on src/App.test.jsx:
```diff
@@ -0,0 +1,26 @@
+
+  const { getByText } = render(<App />);
+
+  expect(getByText('서울')).not.toBeNull();
```
`container` 와 `toHaveTextContent` 를 이용하면 보다 직관적으로 표현할 수 있을 것 같아요

passwd10 commented on src/Regions.test.jsx:
```diff
@@ -0,0 +1,33 @@
+  });
+
+  it('region을 클릭하면 region을 반환한다', () => {
+    const { getByText } = renderElement(sampleRegions);
+    fireEvent.click(getByText('서울'));
+    expect(handleClickRegion).toBeCalledWith({ id: 0, name: '서울' });
```
각각을 공백으로 구분해주시면 문맥 파악이 더 수월해질 것 같아요

passwd10 commented on src/Regions.test.jsx:
```diff
@@ -0,0 +1,33 @@
+    />,
+  );
+
+  it('전달받은 regions를 렌더링 한다.', () => {
```
`전달받은` 이라는 워딩은 불필요해보여요!

passwd10 commented on src/Regions.test.jsx:
```diff
@@ -0,0 +1,33 @@
+    expect(getByText('부산')).not.toBeNull();
+  });
+
+  it('region을 클릭하면 region을 반환한다', () => {
```
테스트 명세와 달리 실제 테스트는 click 이벤트에 대한 테스트를 진행하고 계시네요. `listens click event` 로 표현하면 보다 정확한 명세가 될 것 같아요!

huun0424 replied:
클릭했을 때 ~(V) 가 표시되도록 테스트 명세가 나와있던데, 클릭이벤트 외에 어떤걸 사용하는 것을 의도하셨는지 알 수 있을까요?

passwd10 replied:
아! 제가 말씀드린 명세는 희윤님께서 작성하신 테스트 시나리오를 말한거였어요.

작성하신 테스트 시나리오는 `region을 클릭하면 region을 반환한다` 인데, 실제 테스트에서 확인하고 계신건 click event라 이 부분을 말씀드린거였습니다!

passwd10 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,13 @@
+import { getRegions } from './actions';
+
+describe('reducer', () => {
+  context('getRegions', () => {
```
context는 상반된 상황에서만 선택적으로 사용해야해요. 예를들면 `with task` vs `without task` 같은 상황에서요.

describe로 작성하시면 될 것 같아요!

passwd10 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,13 @@
+
+describe('reducer', () => {
+  context('getRegions', () => {
+    it('regions의 리스트를 가져온다.', () => {
```
reducer가 실제로 수행하는 동작은 regions를 반환하는 일일 거에요.

passwd10 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,13 @@
+describe('reducer', () => {
+  context('getRegions', () => {
+    it('regions의 리스트를 가져온다.', () => {
+      const prevState = { ...initialState };
```
구조분해할당을 하신 이유를 알 수 있을까요?

huun0424 replied:
불변성을 유지하기 위해서 입니다. const prevState = initialState; 로 하면 하위에서 prevState.region = 'ABC'; 등과 같이 객체 내부에 할당을 하게 되면 initialState 값도 함께 바뀌는데, 실수로라도 이런 경우를 방지하기 위해서 값을 참조할땐 deepCopy하도록 습관화 되어있어요.

passwd10 reviewed (commented):
커밋의 단위가 너무 큰 것 같아요 😅 
하나의 커밋에는 하나의 작업만 들어갈 수 있게 해주시고, 의도가 드러나도록 최대한 구체적으로 커밋메세지를 작성해주세요!
그렇게 했을 때 유지보수가 편해질거에요 :)

Commit:
[leehuun][react-week5-assignment]

 - 피드백 반영

Commit:
[leehuun][react-week5-assignment-1]

 - 잘못 커밋된 내용 수정.

Commit:
[leehuun][react-week5-assignment1]

 - setRegion action, reducer, reducer test 추가.
 - Regions comp click event, test 추가.

Commit:
[leehuun][react-week5-assignment-1]

 - test용 sampleData 추가.

Commit:
[leehuun][react-week5-assignment-1]

 - ACTION_KEYS 추가.
 - getCategories, setCategory action, reducer, test 추가.

Commit:
[leehuun][react-week5-assignment-1]

 - Regions comp 예외처리 추가.
 - CategoriesContainer comp, test 추가.
 - Categories comp, test 추가.

passwd10 commented on src/App.test.jsx:
```diff
@@ -6,21 +6,23 @@ import App from './App';
-  const dispatch = jest.fn();
-  useDispatch.mockImplementation(() => dispatch);
+    const dispatch = jest.fn();
+    useDispatch.mockImplementation(() => dispatch);
 
-  useSelector.mockImplementation((selector) => selector({ regions }));
+    useSelector.mockImplementation((selector) => selector({ regions }));
```
dispatch, usedispatch, useSelector 는 describe 바로 하위에 선언해서 사용해주시면 될 것 같아요!

passwd10 commented on src/Regions.jsx:
```diff
@@ -1,13 +1,17 @@
-const Regions = ({ regions, onClickRegion }) => (
+const buttonStyle = { width: '60px', margin: '0 5px' };
+
+const Regions = ({ region, regions, onClickRegion }) => (
```
function 으로 선언한 컴포넌트가있고, arrow function으로 선언한 컴포넌트가 있는데 통일시켜주시면 좋을 것 같아요 :)

passwd10 commented on src/RegionsContainer.jsx:
```diff
@@ -15,9 +17,9 @@ export default function RegionsContainer() {
   }, []);
```
일반적으로 useEffect 는 컴포넌트 가장 하단에 위치시키는편이에요!

passwd10 commented on src/RegionsContainer.jsx:
```diff
@@ -15,9 +17,9 @@ export default function RegionsContainer() {
 
   return (
-    <Regions regions={regions} onClickRegion={handleClickRegion} />
+    <Regions region={region} regions={regions} onClickRegion={handleClickRegion} />
```
```javascript
  <Regions
    region={region}
    regions={regions}
    onClickRegion={handleClickRegion}
  />
```

props가 2개 이상일 때는 세로로 정렬했을 때 가독성이 좋아지는 것 같아요.

passwd10 commented on src/reducer.test.js:
```diff
@@ -12,4 +12,16 @@ describe('reducer', () => {
+
+      const state = reducer(prevState, setRegion(region));
+
+      expect(state.region).not.toBeNull();
```
바로 아래에서 region값을 확인하고 있기 때문에 불필요한 테스트 같아요

passwd10 commented on src/sampleData.js:
```diff
@@ -0,0 +1,27 @@
+export const regionsSample = [
```
일반적으로 mock 데이터들은 `fixtures` 라는 폴더를 만들어서 하위에서 관리해주는 편이에요!

passwd10 commented on src/actions.js:
```diff
@@ -1,6 +1,14 @@
-export const getRegions = () => ({ type: 'getRegions' });
+export const ACTION_KEYS = {
```
action type 을 따로 관리해주고 계시네요 👍 

보통 대문자 snakecase로 네이밍을 하는 편이에요!

https://redux.js.org/usage/reducing-boilerplate#generating-action-creators

passwd10 commented on src/reducer.js:
```diff
@@ -8,7 +9,7 @@ export const initialState = {
 };
 
 const reducer = (state = initialState, action) => {
```
reducer 를 보면 if 가 반복되는 형태에요. action이 많아지면 조건문이 많아져서 복잡도가 올라가겠죠? 어떻게 하면 if문, switch문을 사용하지 않고 문제를 해결할 수 있을까요?

passwd10 commented on src/Categories.test.jsx:
```diff
@@ -0,0 +1,35 @@
+
+  it('categories를 렌더링 한다.', () => {
+    const { getByText } = renderElement(sampleCategories);
+    expect(getByText('한식')).not.toBeNull();
```
```javascript
expect(container).toHaveTextContent('한식');
```

다음과 같이 작성하면 더 직관적으로 테스트를 확인할 수 있을 것 같아요!
